connect-inject/container_init.go

```diff
@@ -72,11 +72,19 @@ func (h *Handler) containerInit(pod corev1.Pod, k8sNamespace string) (corev1.Con
 	}
 
 	// When ACLs are enabled, the ACL token returned from `consul login` is only
-	// valid for a service with the same name as the ServiceAccountName.
-	if data.AuthMethod != "" && data.ServiceName != pod.Spec.ServiceAccountName {
+	// valid for a service with the same name as the ServiceAccountName. If the
+	// annotation for service name is set, check against that.
+	if data.AuthMethod != "" && data.ServiceName != "" && data.ServiceName != pod.Spec.ServiceAccountName {
 		return corev1.Container{}, fmt.Errorf("serviceAccountName %q does not match service name %q", pod.Spec.ServiceAccountName, data.ServiceName)
 	}
```

Comment:
I'm curious if you think it'd make sense to consolidate it and fail in both cases in the init container. That way we'd have a consistent UX for this use-case

Reply:
So I've just been trying it out, and I like the idea, since it requires a lot less comments to understand and one additional flag for the connect-init command. Will modify this PR shortly with the changes

Reply:
I went ahead and had both cases fail in the init container